src/mqtt_client.rs

```diff
@@ -536,8 +571,47 @@ impl<
                         self.network.send_packet(&puback)?;
                     }
 
-                    // TODO: Add support.
-                    QoS::ExactlyOnce => unimplemented!(),
+                    QoS::ExactlyOnce => {
+                        // Note(uwnrap): There should always be a packet ID for QoS >
+                        // AtMostOnce.
+                        let packet_id = info.packet_id.unwrap();
+
+                        // Check if the packet ID already exists before forwarding to app
```

Comment:
Is it certain that the broker will PubRel also the duplicate packet_ids in pending_packets?

Reply:
The broker is required to PubRel the packet_id after receiving the PubRec in order to release the ID for further use. If the broker were to reuse the ID without a preceding PubRel, it would be a protocol violation.

Reply:
Then I don't get the intention behind the handling of duplicate packet_ids in pending_packets here.

Reply:
The broker may have disconnected before it received our PubRec and we're maintaining the session state across a reconnection. This is taken directly from MQTT 4.3.3 and 4.4

Reply:
Ack. Might be good idea to ref the spec in the code where possible.